### PR TITLE
MCR-5335 Delete oauth client

### PR DESCRIPTION
**Pull request by shamoons**
## Summary

This pull request introduces enhancements and new functionality to the OAuth client management system, including the ability to delete OAuth clients, improvements to fetching clients, and updates to the GraphQL schema and resolvers. The changes also include additional tests to ensure robustness.

#### OAuth Client Deletion

- Added a new resolver `deleteOauthClientResolver` to handle the deletion of OAuth clients, restricted to admin users.
- Updated the `Store` in `postgresStore.ts` to include `deleteOAuthClient` functionality.
- Introduced a GraphQL mutation `deleteOauthClient` with the corresponding input and payload types in the schema.

#### Fetching OAuth Clients

- Enhanced the `fetchOauthClients` resolver to handle cases where input is omitted or empty, fetching all clients by default.
- Updated tests for `fetchOauthClients` to cover new scenarios, including fetching with empty input and handling non-existent clients.

#### GraphQL Updates

- Added a new mutation `createOauthClient` to the GraphQL schema and its corresponding implementation.
- Updated the `grants` field in `CreateOauthClientInput` to make it optional.

### Testing Improvements

- Added unit tests for creating and deleting OAuth clients, including edge cases like database failures and unauthorized access.
- Improved existing tests for fetching OAuth clients to ensure comprehensive coverage.

#### Related issues

[_N/A_](https://jiraent.cms.gov/browse/MCR-5335)

#### Screenshots

_N/A_

#### Test cases covered

- `deleteOauthClientResolver.test.ts`:
  - Deletes an OAuth client with admin privileges
  - Handles unauthorized user errors
  - Validates behavior when client ID does not exist
- `fetchOauthClients.test.ts`:
  - Fetches all clients when no input is provided
  - Handles invalid or non-existent client IDs
- `createOauthClient.test.ts`:
  - Successfully creates clients with and without optional fields
  - Handles DB error scenarios and permission restrictions

### CODE REVIEW

File: services/app-api/src/resolvers/oauthClient/deleteOauthClient.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+): MutationResolvers['deleteOauthClient'] {
+    return async (
+        _parent: unknown,
+        args: { input: MutationResolvers['DeleteOauthClientInput'] },
+        context: Context
+    ) => {
+        const { user } = context
+        const { input } = args
+        if (!user || user.role !== 'ADMIN_USER') {
+            throw new Error(
```
Comment (JasonLin0991) [inline services/app-api/src/resolvers/oauthClient/deleteOauthClient.ts:16]:
```suggestion
            const message = 'user not authorized to delete OAuth clients'
            logError('deleteOauthClient', message)
            setErrorAttributesOnActiveSpan(message, span)
            throw new ForbiddenError(message)
```

File: services/app-graphql/src/schema.graphql [hunk 2]
```diff
@@ -2568,3 +2570,12 @@ input FetchOauthClientsInput {
 type FetchOauthClientsPayload {
     oauthClients: [OauthClient!]!
 }
+
+input DeleteOauthClientInput {
+    id: ID
```
Comment (JasonLin0991) [inline services/app-graphql/src/schema.graphql:2575]: **Suggestion:** Lets pick one of these inputs and not have 2. Then make that single input required. I would prefer `clientId` to align with `createOauthClient` mutation.

File: services/app-api/src/resolvers/oauthClient/deleteOauthClient.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+                'Forbidden: Only ADMIN users can delete OAuth clients'
+            )
+        }
+        // Must provide id or clientId
+        if (!input.id && !input.clientId) {
```
Comment (JasonLin0991) [inline services/app-api/src/resolvers/oauthClient/deleteOauthClient.ts:21]: **Suggestion:** If we change the mutation input into one required field, either `id` or `clientIid`, then we won't need this check.

File: services/app-api/src/resolvers/oauthClient/deleteOauthClient.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            throw new Error('Must provide id or clientId')
+        }
+        // Fetch the client
+        let oauthClient
+        if (input.id) {
```
Comment (JasonLin0991) [inline services/app-api/src/resolvers/oauthClient/deleteOauthClient.ts:26]: **Suggestion:** If we change the mutation input into one required field, either `id` or `clientIid`, then we won't need this check.

File: services/app-api/src/resolvers/oauthClient/deleteOauthClient.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+            oauthClient = await store.getOAuthClientById(input.id)
+        } else if (input.clientId) {
+            oauthClient = await store.getOAuthClientByClientId(input.clientId)
+        }
+        if (!oauthClient || oauthClient instanceof Error) {
+            throw new Error('OAuth client not found')
```
Comment (JasonLin0991) [inline services/app-api/src/resolvers/oauthClient/deleteOauthClient.ts:32]:
```suggestion
               const errMessage = `Issue finding OAuth client with clientId ${input.clientId}. ${oauthClient.message}`
               logError('deleteOauthClient', errMessage)
               setErrorAttributesOnActiveSpan(errMessage, span)
               throw new GraphQLError(errMessage, {
                    extensions: {
                        code: 'NOT_FOUND',
                        cause: 'DB_ERROR',
                    },
                })
```

File: services/app-api/src/resolvers/oauthClient/deleteOauthClient.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        }
+        // TODO: Call lambda to deactivate token if needed
+        // Delete from DB
+        const deleted = await store.deleteOAuthClient(oauthClient.id)
+        if (!deleted || deleted instanceof Error) {
+            throw new Error('Failed to delete OAuth client')
```
Comment (JasonLin0991) [inline services/app-api/src/resolvers/oauthClient/deleteOauthClient.ts:38]:
```suggestion
            const errMessage = `Issue deleting OAuth client with clientId ${input.clientId}. ${deleted.message}`
            logError('deleteOauthClient', errMessage)
            setErrorAttributesOnActiveSpan(errMessage, span)
            throw new GraphQLError(errMessage, {
                extensions: {
                    code: 'INTERNAL_SERVER_ERROR',
                    cause: 'DB_ERROR',
                },
            })
```

File: services/app-api/src/resolvers/oauthClient/deleteOauthClient.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        }
+        return { oauthClient: deleted }
```
Comment (JasonLin0991) [inline services/app-api/src/resolvers/oauthClient/deleteOauthClient.ts:40]:
```suggestion

        logSuccess('deleteOauthClient')
        setSuccessAttributesOnActiveSpan(span)
        
        return { oauthClient: deleted }
```

File: services/app-graphql/src/schema.graphql [hunk 8]
```diff
@@ -2568,3 +2570,12 @@ input FetchOauthClientsInput {
...
+    clientId: String
```
Comment (JasonLin0991) [inline services/app-graphql/src/schema.graphql:2576]:
```suggestion
    clientId: String!
```

File: services/app-graphql/src/schema.graphql (hunk 2 above)
Comment (JasonLin0991) [inline services/app-graphql/src/schema.graphql:2575]:
```suggestion
```

File: services/app-api/src/postgres/postgresStore.ts [hunk 9]
```diff
@@ -246,6 +247,7 @@ type Store = {
     getOAuthClientByClientId: (
         clientId: string
     ) => ReturnType<typeof _getOAuthClientByClientId>
+    deleteOAuthClient: (id: string) => ReturnType<typeof _deleteOAuthClient>
 
     prismaClient: ExtendedPrismaClient
```
Comment (JasonLin0991) [inline services/app-api/src/postgres/postgresStore.ts:252]:
```suggestion
```

File: services/app-api/src/postgres/postgresStore.ts [hunk 10]
```diff
@@ -350,6 +352,7 @@ function NewPostgresStore(client: ExtendedPrismaClient): Store {
         getOAuthClientById: (id) => _getOAuthClientById(client, id),
         getOAuthClientByClientId: (clientId) =>
             _getOAuthClientByClientId(client, clientId),
+        deleteOAuthClient: (id) => _deleteOAuthClient(client, id),
 
         prismaClient: client,
```
Comment (JasonLin0991) [inline services/app-api/src/postgres/postgresStore.ts:357]:
```suggestion
```

**Review by JasonLin0991 — APPROVED** [review]
Looks good.

Pretty much the same things as the create OAuth client PR. Great work!
